amy/workshops/models.py

```diff
@@ -2671,6 +2671,34 @@ def clean_event(
         elif not requirement.event_required and self.event:
             return self.get_not_required_error(requirement)
 
+        if self.event and hasattr(self, "trainee") and self.trainee:
+            # check this trainee has a learner task for this event
+            try:
+                Task.objects.get(
+                    person=self.trainee,
+                    event=self.event,
+                    role=Role.objects.get(name="learner"),
+                )
+            except Task.DoesNotExist:
+                msg = (
+                    "This progress cannot be created without a corresponding learner "
+                    f"task. Trainee {self.trainee} does not have a learner task for "
+                    f"event {self.event}."
+                )
+                return ValidationError(msg)
+
+            # check this trainee doesn't already have a training progress for this event
+            return self.clean_event_existing_progress(self.event, self.trainee)
+
+    def clean_event_existing_progress(self, event, trainee):
+        if existing_progress := trainee.trainingprogress_set.filter(event=event):
+            if not (existing_progress.count() == 1 and self in existing_progress):
+                msg = (
+                    f"Trainee {self.trainee} already has a training progress "
+                    f"for event {self.event}."
+                )
+                return ValidationError(msg)
```

Comment:
Could we have a UNIQUE constraint on `(trainee, event)` pair that would move this check to database level?

One thing that should be checked is how the DB treats pairs containing NULL values for `event`.